test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperDropDownListExtensionsTest.cs

```diff
@@ -0,0 +1,256 @@
+            helper.ViewContext.ClientValidationEnabled = false;
+            helper.ViewData.ModelState.SetModelValue("Property1", 1, "1");
+            helper.ViewData["Property1"] = BasicSelectList;
+
+            // Act
+            var dropDownListResult = helper.DropDownList("Property1", "--select--");
+
+            // Assert
+            Assert.Equal(expectedHtml, HtmlContentUtilities.HtmlContentToString(dropDownListResult));
+        }
+
+        [Fact]
+        public void DropDownList_UsesSpecifiedExpressionAndSelectList()
+        {
+            // Arrange
+            var expectedHtml = "<select id=\"HtmlEncode[[Property2]]\" name=\"HtmlEncode[[Property2]]\">" +
+                "<option selected=\"HtmlEncode[[selected]]\" value=\"HtmlEncode[[4]]\">HtmlEncode[[Four]]</option>" + Environment.NewLine +
+                "<option value=\"HtmlEncode[[5]]\">HtmlEncode[[Five]]</option>" + Environment.NewLine +
+                "</select>";
+            var selectList = new List<SelectListItem>
+            {
+                new SelectListItem { Text = "Four", Value = "4" },
+                new SelectListItem { Text = "Five", Value = "5" },
+            };
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(new ViewDataDictionary<TestModel>(metadataProvider));
+            helper.ViewContext.ClientValidationEnabled = false;
+            helper.ViewData.Model = new TestModel { Property2 = "4" };
+
+            // Act
+            var dropDownListResult = helper.DropDownList("Property2", selectList);
+
+            // Assert
+            Assert.Equal(expectedHtml, HtmlContentUtilities.HtmlContentToString(dropDownListResult));
+        }
+
+        [Fact]
+        public void DropDownList_UsesSpecifiedExpressionAndSelectListAndHtmlAttributes()
+        {
+            // Arrange
+            var expectedHtml = "<select id=\"HtmlEncode[[Property1]]\" Key=\"HtmlEncode[[Value]]\" name=\"HtmlEncode[[Property1]]\">" +
+                "<option selected=\"HtmlEncode[[selected]]\" value=\"HtmlEncode[[4]]\">HtmlEncode[[Four]]</option>" + Environment.NewLine +
+                "<option value=\"HtmlEncode[[5]]\">HtmlEncode[[Five]]</option>" + Environment.NewLine +
+                "</select>";
+            var selectList = new List<SelectListItem>
+            {
+                new SelectListItem { Text = "Four", Value = "4" },
+                new SelectListItem { Text = "Five", Value = "5" },
+            };
+            var metadataProvider = new EmptyModelMetadataProvider();
+            var helper = DefaultTemplatesUtilities.GetHtmlHelper(new ViewDataDictionary<TestModel>(metadataProvider));
+            helper.ViewContext.ClientValidationEnabled = false;
+            helper.ViewData.ModelState.SetModelValue("Property1", 4, "4");
+            helper.ViewData["Property1"] = BasicSelectList;
```

Comment:
This is to show that the passed in select list is preferred over the one in `ViewData` as opposed to the case in `DropDownListFor_NullSelectListFindsListFromViewData`.